packages/zone/src/durable.js

```diff
@@ -55,31 +74,49 @@ export const detachedDurableStores = attachDurableStores(() =>
  * Create a zone whose objects persist between Agoric vat upgrades.
  *
  * @param {import('@agoric/vat-data').Baggage} baggage
+ * @param {string} [baseLabel]
  * @returns {import('.').Zone}
  */
-export const makeDurableZone = baggage => {
+export const makeDurableZone = (baggage, baseLabel = 'durableZone') => {
   baggage || Fail`baggage required`;
+
+  const attachedStores = attachDurableStores(() => baggage);
+
+  const { makeOnce, wrapProvider } = makeOnceKit(
+    baseLabel,
+    attachedStores,
+    baggage,
+  );
+
   /** @type {import('.').Zone['exoClass']} */
   const exoClass = (...args) => prepareExoClass(baggage, ...args);
   /** @type {import('.').Zone['exoClassKit']} */
   const exoClassKit = (...args) => prepareExoClassKit(baggage, ...args);
   /** @type {import('.').Zone['exo']} */
   const exo = (...args) => prepareExo(baggage, ...args);
 
-  const attachedStores = attachDurableStores(() => baggage);
+  const mapStore = wrapProvider(attachedStores.mapStore);
 
   /** @type {import('.').Zone['subZone']} */
   const subZone = (label, options = {}) => {
-    const subBaggage = provideDurableMapStore(baggage, label, options);
-    return makeDurableZone(subBaggage);
+    const subBaggage = mapStore(label, options);
+    return makeDurableZone(subBaggage, `${baseLabel}.${label}`);
   };
 
   return Far('durableZone', {
-    exo,
-    exoClass,
-    exoClassKit,
+    exo: wrapProvider(exo, keys.exo),
+    exoClass: wrapProvider(exoClass, keys.exoClass),
+    exoClassKit: wrapProvider(exoClassKit, keys.exoClassKit),
```

Comment:
From what I understand, this change is motivated to allow switching from the current provide pattern to the zone mechanism? However doesn't it introduce a problem for upgrading vats which already use the zone mechanism, if the zone implementation changes behavior?

Reply:
I'm not sure I understand your concern.  The `durable`, `heap`, and `virtual` zones are compatible with `@agoric/vat-data`.  `@agoric/vat-data` cannot change the keys it uses, or else it can't upgrade.  Likewise, those 3 zones cannot change the keys they use, or `durable` cannot be upgraded or `virtual` and `heap` will be incompatible with `durable`.

If we want to break compatibility with the existing `@agoric/vat-data` key names, we should create a new family of zones which are imported differently.

Reply:
> doesn't it introduce a problem for upgrading vats which already use the zone mechanism

The existing zone mechanism already uses the provide pattern.  This just adds enforcement of the underlying key namespaces to it.

Reply:
Resolving